adr/query_deduplication.tla

```diff
@@ -0,0 +1,501 @@
+    wait_map = [ i \in Keys |-> {} ];
+    channels = [i \in Proc |-> FALSE ];
+    tested_keys = {};
+ 
+   define {
+
+    MutualExclusion == \A i, j \in Proc : 
+                         (i # j) => ~ /\ pc[i] = "cs"
+                                      /\ pc[j] = "cs"
+                                      
+    DeadlockFreedom == 
+        \A i \in Proc : 
+          (pc[i] = "exchange") ~> (\E j \in Proc : pc[j] = "cs")
+    
+    
+    \*AllFinished == <>(\A proc \in Proc: pc[proc] = "Done")
+    \*AllKeysLoaded == <>(\A k \in tested_keys: cache[k])
+  };
+
+  procedure lock(idx = 0)
+  {
+    exchange:
+      await ~locks[idx];
+    check:
+      if(~locks[idx]) {
+        locks[idx] := TRUE;
+        return;
+      } else {
+        goto exchange;
+      };
+  };
+
+  procedure unlock(idx = 0)
+  {
+    reset_state:
+      locks[idx] := FALSE;
+      return;
+  }
+  
+  fair process (P \in Proc)
+  variables Key \in Keys, pr = 0;
+  {
+  
+ 
+    start:
+        \* let mut locked_cache = self.cached.lock().await;
+        call lock(cache_lock);
+
+    cache_locked:
+        \*if let Some(value) = locked_cache.get(&key).cloned() { return value }
+        if (cache[Key]) {
+            \* if there is a value for the key, we test two possible behaviours
+            \* expire the key, or assume a normal task that will just return after
+            \* receiving the value
+            either {
+                \* expire one key
+                expiration_lock_1:
+                    cache[Key] := FALSE;
+                    call unlock(cache_lock);
+
+                expiration_done:
+                    goto finished;
+            }
+            or {
+                \*if let Some(value) = locked_cache.get(&key).cloned() { return value }
+                call unlock(cache_lock);
+                cache_hit:
+                    goto finished;
+            }
+        };
+
+    cache_miss:
+        \* we test at the end if all keys have been loaded
+        tested_keys := tested_keys \union {Key};
+    
+    lock_2:
+        \* let mut locked_wait_map = self.wait_map.lock().await;
+        call lock(wait_map_lock);
+    
+    lock_3:
+        \* drop(locked_cache);
+        call unlock(cache_lock);
+    
+    lookup_wait_map:
+        \* match locked_wait_map.get_mut(&key) {
+        \*    Some(waiter) => {
+        if (Cardinality(wait_map[Key]) > 0)
+        {
+            wait_map_subscribe:
+                \* we register to reeive a notification when the value is available
+                \* let mut receiver = waiter.subscribe();
+                wait_map[Key] := wait_map[Key] \union {self};
+                \* drop(locked_wait_map);
+                call unlock(wait_map_lock);
+            
+            waiting_for_value:
+                \* let (recv_key, recv_value) = receiver.recv().await
+                await channels[self];
+                goto finished;
+    
+        }
+        else
+        {
+            wait_map_register:
```

Comment:
if the wait map returns an empty set for the key, that means our process is the first one requesting the key.
I put assertions there that the wait map did not change (it's under the lock, should be fine) and that the cache was still not filled when we reached that state.
We insert our process id in the wait map then unlock it. The `fetching_value` step contains a `skip` instruction, which will arbitrarily pause the process here and let others advance. It represents the work we are trying to cache (query planning, etc)